src/backend/distributed/transaction/multi_transaction.c

```diff
@@ -0,0 +1,1907 @@
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	bool found = false;
+
+	/* There seems to be little point in making this work */
+	Assert(InCoordinatedTransaction());
+
+	placementKey.placementid = placement->placementId;
+
+	placementEntry = hash_search(ConnectionPlacementHash, &placementKey, HASH_ENTER, &found);
+	if (!found)
+	{
+		placementEntry->connections = NIL;
+		placementEntry->required = false;
+	}
+
+	placementEntry->failed = true;
+
+	AssociatePlacementShard(placementEntry, placement);
+}
+
+
+void
+MarkPlacementConnectionRequired(ShardPlacement *placement)
+{
+	ConnectionPlacementHashKey placementKey;
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	bool found = false;
+
+	/* There seems to be little point in making this work */
+	Assert(InCoordinatedTransaction());
+
+	placementKey.placementid = placement->placementId;
+
+	placementEntry = hash_search(ConnectionPlacementHash, &placementKey, HASH_ENTER, &found);
+	if (!found)
+	{
+		placementEntry->connections = NIL;
+		placementEntry->failed = false;
+	}
+
+	placementEntry->required = true;
+
+	AssociatePlacementShard(placementEntry, placement);
+}
+
+
+static void
+AssociatePlacementShard(ConnectionPlacementHashEntry *placementEntry, ShardPlacement *placement)
+{
+	ConnectionShardHashKey shardKey;
+	ConnectionShardHashEntry *shardEntry = NULL;
+	bool found = false;
+	MemoryContext oldContext = NULL;
+
+	shardKey.shardId = placement->shardId;
+	shardEntry = hash_search(ConnectionShardHash, &shardKey, HASH_ENTER, &found);
+	if (!found)
+	{
+		shardEntry->placementConnections = NIL;
+	}
+
+	/* XXX: has to be unique, right? */
+	oldContext = MemoryContextSwitchTo(CacheMemoryContext);
+	shardEntry->placementConnections =
+		list_append_unique_ptr(shardEntry->placementConnections, placementEntry);
+	MemoryContextSwitchTo(oldContext);
+}
+
+
+extern MultiConnection *
+StartPlacementConnection(uint32 flags, ShardPlacement *placement)
+{
+	ConnectionPlacementHashKey key;
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	MultiConnection *connection;
+	MemoryContext oldContext;
+	bool found;
+
+	key.placementid = placement->placementId;
+
+	/* FIXME: not implemented */
+	Assert(flags & NEW_CONNECTION);
+
+	/*
+	 * Lookup relevant hash entry. We always enter. If only a cached
+	 * connection is desired, and there's none, we'll simply leave the
+	 * connection list empty.
+	 */
+
+	placementEntry = hash_search(ConnectionPlacementHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		placementEntry->connections = NIL;
+		placementEntry->failed = false;
+		placementEntry->required = false;
+	}
+
+	/* XXX: do we want to error out if placement is in failed state? */
+
+	if (flags & CACHED_CONNECTION)
+	{
+		ListCell *connectionCell = NULL;
+
+		/* check connection cache for a connection that's not already in use */
+		foreach(connectionCell, placementEntry->connections)
+		{
+			connection = (MultiConnection *) lfirst(connectionCell);
+
+			if (!connection->busy)
```

Comment:
On 2016-09-08 06:42:47 -0700, Marco Slot wrote:
> > +		placementEntry->failed = false;
> > +		placementEntry->required = false;
> > +	}
> > +
> > +	/* XXX: do we want to error out if placement is in failed state? */
> > +
> > +	if (flags & CACHED_CONNECTION)
> > +	{
> > +		ListCell *connectionCell = NULL;
> > +
> > +		/* check connection cache for a connection that's not already in use */
> > +		foreach(connectionCell, placementEntry->connections)
> > +		{
> > +			connection = (MultiConnection *) lfirst(connectionCell);
> > +
> > +			if (!connection->busy)
>
> Is this flag a workaround for opening one connection per placement?

Well, it's there because a connection to a remote node, which can be
shared between placements, might already be busy doing something
else. That'll e.g. frequently be the case if you first do a bunch of
router selects and then do COPY or such.


> Maybe add an explicit CONNECTION_PER_PLACEMENT flag instead. The caller
> probably wants to use the cached connection, because it may have
> modified the placement, but if that connection is still in use then the
> caller made a mistake somewhere and we should error out or
> Assert(!connection->busy).

I am thinking of doing that a bit differently. Because a connection can,
but does not have to, be used for multiple placements, we basically need
to check that for all the placements involved in the query, the returned
connection has to be the only dirty one.  That'll then recognize the
troubling case of COPY opening per placement connections, followed by a
(router) query involving different placements.  For that we need an API
that specifies all the placements that a connection is going to be used
for.

I think that's primarily going to be relevant for router and real-time,
but maybe there's other possible users?